sympy/matrices/tests/test_decompositions.py

```diff
@@ -91,6 +91,35 @@ def test_LUdecomp():
     )
     raises(ValueError, lambda : M.LUdecomposition_Simple(rankcheck=True))
 
+def test_SVD():
+    A = Matrix([[1, 2], [2, 1]])
+    U, S, V = A.SVD()
+    assert U * S * V.T == A
+    assert U.T * U == eye(U.cols)
+    assert V.T * V == eye(V.cols)
+
+    B = Matrix([[1, 2]])
+    U, S, V = B.SVD()
+
+    assert U * S * V.T == B
+    assert U.T * U == eye(U.cols)
+    assert V.T * V == eye(V.cols)
+
+    C = Matrix([
+        [1, 0, 0, 0, 2],
+        [0, 0, 3, 0, 0],
+        [0, 0, 0, 0, 0],
+        [0, 2, 0, 0, 0],
+    ])
+
+    U, S, V = C.SVD()
+
+    assert U * S * V.T == C
+    assert U.T * U == eye(U.cols)
+    assert V.T * V == eye(V.cols)
```

Comment:
Just a passing commit, seems like this lacks tests for fully symbolic matrices.

Reply:
That's a good suggestion. I'll look into it

Reply:
@moorepants @sylee957 I tested on a few symbolic matrices like the example below and as expected there is an expression blow-up. I don't think this issue can be resolved because of the way SVD and diagonalization algorithms work for symbolic matrices. I am open to any suggestions to help fix this


>>> from sympy import Matrix,sqrt,Rational
>>> A = Matrix([[Rational(1/3), sqrt(2)],[0, Rational(1,4)]])
>>> U, S, V = A.SVD()
>>> U
Matrix([
[18014398509481984*sqrt(6)*(6004799503160661*(-sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232 - 211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232)/(18014398509481984*sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2)) + sqrt(2)/sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2))/(3*sqrt(235125711331221215061729356798483 - sqrt(54959062191764135232862476524808863960980457719206728824672116073))), 18014398509481984*sqrt(6)*(6004799503160661*(-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)/(18014398509481984*sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1)) + sqrt(2)/sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1))/(3*sqrt(sqrt(54959062191764135232862476524808863960980457719206728824672116073) + 235125711331221215061729356798483))],
[                                                                                                                                                                                                                                                                                                                                                                                                                                                      4503599627370496*sqrt(6)/(3*sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2)*sqrt(235125711331221215061729356798483 - sqrt(54959062191764135232862476524808863960980457719206728824672116073))),                                                                                                                                                                                                                                                                                                                                                                                                                                                        4503599627370496*sqrt(6)/(3*sqrt(sqrt(54959062191764135232862476524808863960980457719206728824672116073) + 235125711331221215061729356798483)*sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1))]])
>>> S
Matrix([
[sqrt(705377133993663645185188070395449/649037107316853453566312041152512 - 3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512),                                                                                                                                                                                       0],
[                                                                                                                                                                                      0, sqrt(3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512 + 705377133993663645185188070395449/649037107316853453566312041152512)]])
>>> U * S * V.H 
Matrix([
[18014398509481984*sqrt(6)*sqrt(705377133993663645185188070395449/649037107316853453566312041152512 - 3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512)*(-sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232 - 211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232)*(6004799503160661*(-sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232 - 211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232)/(18014398509481984*sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2)) + sqrt(2)/sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2))/(3*sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2)*sqrt(235125711331221215061729356798483 - sqrt(54959062191764135232862476524808863960980457719206728824672116073))) + 18014398509481984*sqrt(6)*(-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)*sqrt(3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512 + 705377133993663645185188070395449/649037107316853453566312041152512)*(6004799503160661*(-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)/(18014398509481984*sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1)) + sqrt(2)/sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1))/(3*sqrt(sqrt(54959062191764135232862476524808863960980457719206728824672116073) + 235125711331221215061729356798483)*sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1)), 18014398509481984*sqrt(6)*sqrt(705377133993663645185188070395449/649037107316853453566312041152512 - 3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512)*(6004799503160661*(-sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232 - 211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232)/(18014398509481984*sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2)) + sqrt(2)/sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2))/(3*sqrt(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2)*sqrt(235125711331221215061729356798483 - sqrt(54959062191764135232862476524808863960980457719206728824672116073))) + 18014398509481984*sqrt(6)*sqrt(3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512 + 705377133993663645185188070395449/649037107316853453566312041152512)*(6004799503160661*(-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)/(18014398509481984*sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1)) + sqrt(2)/sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1))/(3*sqrt(sqrt(54959062191764135232862476524808863960980457719206728824672116073) + 235125711331221215061729356798483)*sqrt((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1))],
[                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      4503599627370496*sqrt(6)*sqrt(705377133993663645185188070395449/649037107316853453566312041152512 - 3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512)*(-sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232 - 211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232)/(3*(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2)*sqrt(235125711331221215061729356798483 - sqrt(54959062191764135232862476524808863960980457719206728824672116073))) + 4503599627370496*sqrt(6)*(-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)*sqrt(3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512 + 705377133993663645185188070395449/649037107316853453566312041152512)/(3*sqrt(sqrt(54959062191764135232862476524808863960980457719206728824672116073) + 235125711331221215061729356798483)*((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1)),                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       4503599627370496*sqrt(6)*sqrt(705377133993663645185188070395449/649037107316853453566312041152512 - 3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512)/(3*(1 + (211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2)*sqrt(235125711331221215061729356798483 - sqrt(54959062191764135232862476524808863960980457719206728824672116073))) + 4503599627370496*sqrt(6)*sqrt(3*sqrt(54959062191764135232862476524808863960980457719206728824672116073)/649037107316853453566312041152512 + 705377133993663645185188070395449/649037107316853453566312041152512)/(3*sqrt(sqrt(54959062191764135232862476524808863960980457719206728824672116073) + 235125711331221215061729356798483)*((-211087299949115534265110171493869*sqrt(2)/144230468292634092786114449375232 + sqrt(109918124383528270465724953049617727921960915438413457649344232146)/144230468292634092786114449375232)**2 + 1))]])
>>> assert U * S * V.H == A
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError

Reply:
We get an assertion error for the above matrix
assert U * S * V.H == A
Traceback (most recent call last):
File "", line 1, in
AssertionError

Reply:
Usually most of the educational examples for this could be reverse engineered like - similarity transformation of diagonal matrix, and not via random matrix. 

If this is giving complicated results for such textbook examples, this may need to be improved

Reply:
You should use `Rational(1, 3)` or `S(1)/3` rather than `Rational(1/3)`:
```python
In [7]: Matrix([[Rational(S(1)/3), sqrt(2)],[0, Rational(1,4)]])
Out[7]: 
⎡1/3  √2 ⎤
⎢        ⎥
⎣ 0   1/4⎦

In [8]: Matrix([[Rational(1/3), sqrt(2)],[0, Rational(1,4)]])
Out[8]: 
⎡ 6004799503160661     ⎤
⎢─────────────────  √2 ⎥
⎢18014398509481984     ⎥
⎢                      ⎥
⎣        0          1/4⎦
```
Then the result is a lot simpler:
```python
In [9]: U, Si, V = A.SVD()

In [10]: U
Out[10]: 
⎡      ⎛         17⋅√674   281⋅√2                                        ⎞        ⎛          281⋅√2   17⋅√674                                           ⎞⎤
⎢      ⎜       - ─────── - ──────                                        ⎟        ⎜        - ────── + ───────                                           ⎟⎥
⎢      ⎜           192      192                          √2              ⎟        ⎜           192       192                            √2               ⎟⎥
⎢12⋅√2⋅⎜──────────────────────────────── + ──────────────────────────────⎟  12⋅√2⋅⎜────────────────────────────────── + ────────────────────────────────⎟⎥
⎢      ⎜       _________________________        _________________________⎟        ⎜       ___________________________        ___________________________⎟⎥
⎢      ⎜      ╱                       2        ╱                       2 ⎟        ⎜      ╱                     2            ╱                     2     ⎟⎥
⎢      ⎜     ╱      ⎛281⋅√2   17⋅√674⎞        ╱      ⎛281⋅√2   17⋅√674⎞  ⎟        ⎜     ╱  ⎛  281⋅√2   17⋅√674⎞            ╱  ⎛  281⋅√2   17⋅√674⎞      ⎟⎥
⎢      ⎜3⋅  ╱   1 + ⎜────── + ───────⎟       ╱   1 + ⎜────── + ───────⎟  ⎟        ⎜3⋅  ╱   ⎜- ────── + ───────⎟  + 1      ╱   ⎜- ────── + ───────⎟  + 1 ⎟⎥
⎢      ⎝  ╲╱        ⎝ 192       192  ⎠     ╲╱        ⎝ 192       192  ⎠  ⎠        ⎝  ╲╱    ⎝   192       192  ⎠         ╲╱    ⎝   192       192  ⎠      ⎠⎥
⎢─────────────────────────────────────────────────────────────────────────  ─────────────────────────────────────────────────────────────────────────────⎥
⎢                              _______________                                                              _______________                              ⎥
⎢                            ╲╱ 313 - 17⋅√337                                                             ╲╱ 17⋅√337 + 313                               ⎥
⎢                                                                                                                                                        ⎥
⎢                                  3⋅√2                                                                         3⋅√2                                     ⎥
⎢            ────────────────────────────────────────────────                            ──────────────────────────────────────────────────              ⎥
⎢                 _________________________                                                                     ___________________________              ⎥
⎢                ╱                       2                                                                     ╱                     2                   ⎥
⎢               ╱      ⎛281⋅√2   17⋅√674⎞     _______________                              _______________    ╱  ⎛  281⋅√2   17⋅√674⎞                    ⎥
⎢              ╱   1 + ⎜────── + ───────⎟  ⋅╲╱ 313 - 17⋅√337                             ╲╱ 17⋅√337 + 313 ⋅  ╱   ⎜- ────── + ───────⎟  + 1               ⎥
⎣            ╲╱        ⎝ 192       192  ⎠                                                                  ╲╱    ⎝   192       192  ⎠                    ⎦

In [11]: Si
Out[11]: 
⎡    _______________                     ⎤
⎢   ╱ 313   17⋅√337                      ⎥
⎢  ╱  ─── - ───────            0         ⎥
⎢╲╱   288     288                        ⎥
⎢                                        ⎥
⎢                         _______________⎥
⎢                        ╱ 17⋅√337   313 ⎥
⎢         0             ╱  ─────── + ─── ⎥
⎣                     ╲╱     288     288 ⎦

In [12]: V
Out[12]: 
⎡        17⋅√674   281⋅√2                 281⋅√2   17⋅√674       ⎤
⎢      - ─────── - ──────               - ────── + ───────       ⎥
⎢          192      192                    192       192         ⎥
⎢──────────────────────────────  ────────────────────────────────⎥
⎢     _________________________       ___________________________⎥
⎢    ╱                       2       ╱                     2     ⎥
⎢   ╱      ⎛281⋅√2   17⋅√674⎞       ╱  ⎛  281⋅√2   17⋅√674⎞      ⎥
⎢  ╱   1 + ⎜────── + ───────⎟      ╱   ⎜- ────── + ───────⎟  + 1 ⎥
⎢╲╱        ⎝ 192       192  ⎠    ╲╱    ⎝   192       192  ⎠      ⎥
⎢                                                                ⎥
⎢              1                                1                ⎥
⎢──────────────────────────────  ────────────────────────────────⎥
⎢     _________________________       ___________________________⎥
⎢    ╱                       2       ╱                     2     ⎥
⎢   ╱      ⎛281⋅√2   17⋅√674⎞       ╱  ⎛  281⋅√2   17⋅√674⎞      ⎥
⎢  ╱   1 + ⎜────── + ───────⎟      ╱   ⎜- ────── + ───────⎟  + 1 ⎥
⎣╲╱        ⎝ 192       192  ⎠    ╲╱    ⎝   192       192  ⎠      ⎦
```
For the pure algebraic case like this functions like `radsimp` can help:
```
In [23]: U.applyfunc(radsimp).applyfunc(sqrtdenest).applyfunc(factor_terms)
Out[23]: 
⎡  _________________                           _________________                       ⎤
⎢╲╱ 281⋅√337 + 5729 ⋅(-337⋅√34 + 17⋅√11458)  ╲╱ 5729 - 281⋅√337 ⋅(17⋅√11458 + 337⋅√34) ⎥
⎢──────────────────────────────────────────  ───────────────────────────────────────── ⎥
⎢                  91664                                       91664                   ⎥
⎢                                                                                      ⎥
⎢                        _________________     _________________                       ⎥
⎢(337⋅√17 + 119⋅√5729)⋅╲╱ 281⋅√337 + 5729    ╲╱ 5729 - 281⋅√337 ⋅(-337⋅√17 + 119⋅√5729)⎥
⎢─────────────────────────────────────────   ──────────────────────────────────────────⎥
⎣                 1099968                                     1099968                  ⎦
```

Reply:
@oscarbenjamin for this matrix specifically simplify does the job. I have tried some other examples where radsimp proved useful in getting back U*S*V.H == A
>>> A = Matrix([[Rational(1,3), sqrt(2)],[0, Rational(1,4)]])
>>> U, S, V = A.SVD()
>>> simplify(U * S * V.H)
Matrix([
[1/3, sqrt(2)],
[  0,     1/4]])

Reply:
We don't want to use general purpose `simplify` in library code but something more specialised and well defined like `radsimp` could potentially be used as part of the routine for computing the SVD.

Reply:
for some 3 x 3 or higher symbolic matrices the QRdecomposition part of the code stalls 
>>> A = Matrix([[Rational(1,3), sqrt(2), 0],[0, Rational(1,4), 0],[1, 1, sqrt(2)]])
>>> U, S, V = A.SVD()

Reply:
> for some 3 x 3 or higher symbolic matrices the QRdecomposition part of the code stalls
> 
> > > > A = Matrix([[Rational(1,3), sqrt(2), 0],[0, Rational(1,4), 0],[1, 1, sqrt(2)]])
> > > > U, S, V = A.SVD()

In order to solve this I tried using simplify on matrix V / U before the orthogonalization step but that  doesn't help either. I tried using _QRdecomposition_optional(V,normalize=False) so as to remove the normalization step of the orthogonal matrices but even that did not help